rllib/core/rl_module/rl_module.py

```diff
@@ -77,23 +100,106 @@ class RLModule(abc.ABC):
         More details here: https://github.com/pytorch/pytorch/issues/49726.
     """
 
-    @OverrideToImplementCustomLogic_CallToSuperRecommended
-    def __init__(self, config: Mapping[str, Any] = None) -> None:
-        self.config = config or {}
-        self.setup()
+    def __init_subclass__(cls, **kwargs):
+        # Automatically add a __post_init__ method to all subclasses of RLModule.
+        def init_decorator(previous_init):
+            def new_init(self, *args, **kwargs):
+                previous_init(self, *args, **kwargs)
+                if type(self) == cls:
+                    self.__post_init__()
+
+            return new_init
+
+        cls.__init__ = init_decorator(cls.__init__)
+
+    def __post_init__(self):
+        """Called after the __init__ method of the subclass.
```

Comment:
Can you specify that this will be called automatically in all sub-classes, even if the user "forgets" to call it in their own constructors?

Btw, what's the standard here to do these things. I always feel like messing with the user's code in this way is very detrimental and intransparent. We could just do a `init()` and `setup()` method (and both would be called by the base-cass's `__init__` method, which should not be overridden, then.

Reply:
I like this pattern better than `init()` `setup()`. It was introduced in dataclasses as well. It's useful for when you need this base class to get constructed by it's sub-classes but there are parts of the construction that depends on the parameters that user has specify first. 

```
class Base:
     def __init__():
            self.initialize_me()
     
     def initialize_me():
            raise NotImplementedError

# this will break because super() has to be called only after self.foo is defined. 
# so the user has to know this. 
# if initialize_me() is called in the `__post_init__()` 
# method user has more flexibility on where to call `super().__init__()`
class A(Base):
     def __init__():
           super().__init__()
           self.foo = "foo"
      
       def initialize_me():
          if self.foo == "foo":
                  self.bar = "zoo"
          else:
                  self.bar = "goo"
```